src/basic/log.c

```diff
@@ -1131,6 +1131,10 @@ void log_parse_environment_realm(LogRealm realm) {
                  * closed the console yet. */
                 (void) proc_cmdline_parse(parse_proc_cmdline_item, NULL, PROC_CMDLINE_STRIP_RD_PREFIX);
 
+        e = getenv("JOURNAL_STREAM");
```

Comment:
so we set the env var to the st_dev and st_ino fields of the fd stderr is connected to. The idea is that this allows us to protect us from cases where we are forked off some service that itself had stderr connected to the journal but connects our stderr to a pipe it allocated itself, but didn#t clean up $JOURNAL_STREAM.

We should thus compare the contents with the data from fstat(STDERR_FILENO). i.e. add a new helper call or so:

```c
bool stderr_is_journal(void) {
        _cleanup_free_ char *w = NULL;
        const char *e;
        uint64_t dev, ino;

        e = getenv("JOURNAL_STREAM");
        if (!e)
                return false;

        if (extract_first_word(&e, &w, ":",  EXTRACT_DONT_COALESCE_SEPARATORS) <= 0)
                return false;

         if (safe_atou64(w, &dev) < 0)
                return false;
         if (safe_atou64(e, &ino) < 0)
                return false;        

        if (fstat(STDERR_FILENO, &st) < 0)
                return false;

        return st.st_dev == dev & st.st_ino == ino;
}
```

Reply:
i wonder if we should do this unconditionally in log_open() if LOG_TARGET_AUTO is chosen — entirely replacing the isatty() check

Reply:
I replaced the `isatty` check by the `stderr_is_journal` check. I think that's a net improvement over the `isatty` check.

Also, is `extract_many_words` supposed to succeed if it can't read the requested amount of words? That gave me a pretty nasty segfault.